
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Software Features &#8212; PrimerDriver v1.0.0 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Getting Started" href="gettingstarted.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="software-features">
<span id="id1"></span><h1>Software Features<a class="headerlink" href="#software-features" title="Permalink to this headline">¶</a></h1>
<div class="section" id="parameter-setting">
<span id="id2"></span><h2>Parameter Setting<a class="headerlink" href="#parameter-setting" title="Permalink to this headline">¶</a></h2>
<p>The user parameter setting is possible for the inputs in PrimerDriver. The tool designs the desired mutagenic primer pairs according to user input and performs site-directed mutagenesis from single- or multiple-bases within the sequence.</p>
<p>The user needs to open the <code class="docutils literal notranslate"><span class="pre">pdcli/settings.json</span></code> file to define the range of allowable values in most of the parameters. The <code class="docutils literal notranslate"><span class="pre">json</span></code> file with the default settings appears as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span>
    <span class="s2">&quot;Tm_range_min&quot;</span><span class="p">:</span> <span class="mi">75</span><span class="p">,</span>
    <span class="s2">&quot;Tm_range_max&quot;</span><span class="p">:</span> <span class="mi">85</span><span class="p">,</span>
    <span class="s2">&quot;gc_range_min&quot;</span><span class="p">:</span> <span class="mi">40</span><span class="p">,</span>
    <span class="s2">&quot;gc_range_max&quot;</span><span class="p">:</span> <span class="mi">60</span><span class="p">,</span>
    <span class="s2">&quot;length_min&quot;</span><span class="p">:</span> <span class="mi">25</span><span class="p">,</span>
    <span class="s2">&quot;length_max&quot;</span><span class="p">:</span> <span class="mi">45</span><span class="p">,</span>
    <span class="s2">&quot;flank5_range_min&quot;</span><span class="p">:</span> <span class="mi">11</span><span class="p">,</span>
    <span class="s2">&quot;flank5_range_max&quot;</span><span class="p">:</span> <span class="mi">21</span><span class="p">,</span>
    <span class="s2">&quot;flank3_range_min&quot;</span><span class="p">:</span> <span class="mi">11</span><span class="p">,</span>
    <span class="s2">&quot;flank3_range_max&quot;</span><span class="p">:</span> <span class="mi">21</span><span class="p">,</span>
    <span class="s2">&quot;forward_overlap5&quot;</span><span class="p">:</span> <span class="mi">9</span><span class="p">,</span>
    <span class="s2">&quot;forward_overlap3&quot;</span><span class="p">:</span> <span class="mi">9</span><span class="p">,</span>
    <span class="s2">&quot;terminate_gc&quot;</span> <span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
    <span class="s2">&quot;center_mutation&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
    <span class="s2">&quot;primer_mode&quot;</span><span class="p">:</span> <span class="s2">&quot;overlapping&quot;</span><span class="p">,</span>
    <span class="s2">&quot;expression_system&quot;</span><span class="p">:</span> <span class="s2">&quot;Homo sapiens&quot;</span>
<span class="p">}</span>
</pre></div>
</div>
<p>The following definitions and considerations should be made for designing mutagenic primers:</p>
<ul>
<li><dl>
<dt><code class="docutils literal notranslate"><span class="pre">Tm_range</span></code></dt><dd><p>This is where the user sets the minimum and maximum allowable melting temperature (<span class="math notranslate nohighlight">\(T_m\)</span>) of primers in a PCR reaction. Depending on the type of mutation, the melting temperature is calculated differently. Estimation of the Tm of primers for base substitution is defined by</p>
<div class="math notranslate nohighlight">
\[T_m = 81.5 + 0.41(\%\textrm{GC}) - \frac{675}{N} - \%\textrm{mismatch}\]</div>
<p>where <span class="math notranslate nohighlight">\(N\)</span> is the primer length in base pairs, and <span class="math notranslate nohighlight">\(\%\textrm{GC}, \%\textrm{mismatch}\)</span> are whole numbers. A modified formula is used for calculating <span class="math notranslate nohighlight">\(T_m\)</span> that intends to introduce deletions or insertions:</p>
<div class="math notranslate nohighlight">
\[T_m = 81.5 + 0.41(\%\textrm{GC}) - \frac{675}{N}\]</div>
</dd>
</dl>
</li>
<li><dl>
<dt><code class="docutils literal notranslate"><span class="pre">gc_range</span></code></dt><dd><p>The GC content is calculated as</p>
<div class="math notranslate nohighlight">
\[\%\textrm{GC} = \frac{\textrm{# of G + # of C}}{\textrm{total length of primer}}\]</div>
<p>and the range of GC in a primer is to be set by the user. The user can follow the percentage as prescribed by a protocol or the user can just set the range from 0 to 100.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">length</span></code></dt><dd><p>The total length of the desired mutagenic primer</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">flank5_range</span></code></dt><dd><p>This is the length of the region that is to the left of the mutation when read in standard sequence notation.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">flank3_range</span></code></dt><dd><p>This refers to the length of the region that is to the right of the mutation when read in 5’ to 3’.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">forward_overlap5</span></code></dt><dd><p>When generating overlapping primers, this is where the user sets the minimum number of overlapping bases of the 5’ end of the forward primer to the reverse primer.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">forward_overlap3</span></code></dt><dd><p>This sets the minimum number of base overlaps from the 3’ end of the forward primer to the reverse primer.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">terminate_gc</span></code></dt><dd><p>Set in boolean values, this refers to having a GC clamp at both ends of the primers.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">center_mutation</span></code></dt><dd><p>This sets the mutation is centered within the primer. Flanking regions differ by at most one base pair. These are set by boolean values.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">primer_mode</span></code></dt><dd><p>This gives the user the choice of having overlapping primers or complementary primers. Both modes use protocol from QuikChange™ Site-Directed Mutagenesis Kit by Stratagene®.</p>
<ul class="simple">
<li><dl class="simple">
<dt>Complementary</dt><dd><p>The values for the flanking regions will just be applied to the forward primer. The reverse primers will just be the reverse complements of the forward primers.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Overlapping</dt><dd><p>Both the forward and reverse primers will be screened for the flanking region values. Forward overlap values will also be applied when generating the primer pairs.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">expression_system</span></code></dt><dd><p>When using Protein based primer design, the user can choose from a number of commonly used model organisms used in molecular and cell biology. This enables the tool to generate primers that use the optimum codon for the chosen organism.</p>
</dd>
</dl>
</li>
</ul>
</div>
<div class="section" id="input-format">
<span id="id3"></span><h2>Input format<a class="headerlink" href="#input-format" title="Permalink to this headline">¶</a></h2>
<p>In PrimerDriver, sequences can be added either by copying/pasting into the input cursor. The tool also accepts one or multiple sequences when uploaded in FASTA format. The input sequence must match the DNA sequence that will serve as template for PCR. Before processing the sequences, a basic input emending (FASTA headers, unsupported characters) is performed.</p>
</div>
<div class="section" id="choosing-an-interface">
<span id="interface"></span><h2>Choosing an interface<a class="headerlink" href="#choosing-an-interface" title="Permalink to this headline">¶</a></h2>
<p>As a powerful and useful feature in DNA engineering, the Web server offers the possibility of designing primers for site-directed mutagenesis. The tool can accommodate both DNA and protein sequences to incorporate base pair insertions, deletions, and substitutions as specified by the user. This can cater to an array of primer designs for site-directed single mutagenesis. PrimerDriver also lets you choose from two different command-line modes:</p>
<ul class="simple">
<li><dl class="simple">
<dt>Interactive mode</dt><dd><p>Guides the user through a step-by-step input prompt. Primarily aimed towards demonstration purposes and first-time users of the program.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Single-command mode</dt><dd><p>Requires the user to include all the inputs and arguments needed to perform the task in a single command. Primarily aimed towards more advanced tasks such as batch primer design.</p>
</dd>
</dl>
</li>
</ul>
</div>
<div class="section" id="choosing-primer-design-modes">
<span id="design-mode"></span><h2>Choosing primer design modes<a class="headerlink" href="#choosing-primer-design-modes" title="Permalink to this headline">¶</a></h2>
<p>PrimerDriver asks you to choose from three different modes when using the command-line interface:</p>
<blockquote>
<div><ul class="simple">
<li><p>Primer characterization</p></li>
<li><p>DNA-based primer design</p></li>
<li><p>Protein-based primer design</p></li>
</ul>
</div></blockquote>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">PrimerDriver</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="gettingstarted.html">Getting Started</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Software Features</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#parameter-setting">Parameter Setting</a></li>
<li class="toctree-l2"><a class="reference internal" href="#input-format">Input format</a></li>
<li class="toctree-l2"><a class="reference internal" href="#choosing-an-interface">Choosing an interface</a></li>
<li class="toctree-l2"><a class="reference internal" href="#choosing-primer-design-modes">Choosing primer design modes</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/kvdomingo/primerdriver">Source Code</a></li>
<li class="toctree-l1"><a class="reference external" href="https://primerdriver.herokuapp.com/">Official Website</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="gettingstarted.html" title="previous chapter">Getting Started</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, Kenneth Domingo, Nomer Gutierrez, Shebna Fabilloren, and Carlo Lapid.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.4.3</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/software.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>